### PR TITLE
Always prefer causes when backtracking

### PR DESCRIPTION
**Pull request by notatallshaw**
As per https://github.com/sarugaku/resolvelib/issues/131 this moves the responsibility from the downstream library to prefer backtracking causes into resolvelib itself solving the issue of an O(n<sup>2</sup>) performance problem that can happen.

It provides additional performance benefits under those circumstances because `get_preference` will be called a lot less.

After working on this PR I decided not to have this as optional, it keeps the resolver and tests much simpler and while it may be possible to construct a dependency graph where this approach is slower we have no such reports come from Pip and as per [this comment](https://github.com/sarugaku/resolvelib/issues/64#issuecomment-1567302535) in real world benchmarking there is over a 2x improvement in resolution speed to preferring backtracking, this makes intuitive sense as any successful solution must not have conflicts so resolving them first will generally be faster.

But let me know if you would prefer this approach to be optional, I can do so.

I have been testing these change in Pip on this branch: https://github.com/pypa/pip/compare/main...notatallshaw:pip:always-prefer-causes

### CODE REVIEW

File: src/resolvelib/resolvers.py [hunk 1]
```diff
@@ -418,10 +417,29 @@ def resolve(self, requirements, max_rounds):
...
+                )
+                if unsatisfied_causes_names:
+                    unsatisfied_names = list(unsatisfied_causes_names)
+
+                self._r.backtracking_on(
+                    backtrack_cause_names, unsatisfied_causes_names
+                )
+
             # Choose the most preferred unpinned criterion to try.
             name = min(unsatisfied_names, key=self._get_preference)
```
Comment (notatallshaw) [inline src/resolvelib/resolvers.py:444]: I could put a check here to see if the length of unsatisfied names is 1 and skip calling `self._get_preference` altogether under this scenario?

File: src/resolvelib/providers.py [hunk 2]
```diff
@@ -9,14 +9,7 @@ def identify(self, requirement_or_candidate):
...
 
-    def get_preference(
-        self,
-        identifier,
-        resolutions,
-        candidates,
-        information,
-        backtrack_causes,
-    ):
+    def get_preference(self, identifier, resolutions, candidates, information):
```
Comment (uranusjr) [inline src/resolvelib/providers.py:12]: We’re at version 1 now so it’d be better if we could implement this in a backward-compatible way. Maybe we can still pass the backtrack causes anyway?

**Comment by uranusjr** [comment]
Logic looks good, one question about interface change. Would it be possible to have a test for this?

File: src/resolvelib/providers.py (hunk 2 above)
Comment (notatallshaw) [reply src/resolvelib/providers.py:12]: Sure, it was just to simplify the interface and make sure downstream libraries aren't duplicating this logic, but I don't have any strong opinions on whether to keep providing it or not.

**Comment by notatallshaw** [comment]
> Logic looks good, one question about interface change. Would it be possible to have a test for this?

I will look at the existing tests and see if I can sufficiently adapt one.

**Comment by notatallshaw** [comment]
I'm closing this, this optimization is relatively niche in real world (only one reported case of significant slowdown and never been able to reproduce).

But I do have a bigger optimization that will extend this strategey to "always prefer conflicting causes, then causes", I beleive that optimization, so this approach will be included in that and have stronger justification with lots of real world examples I can show it speeds up. I will make a new PR for that one.

**Pull request closed without merging**